src/partialBind.ts

```diff
@@ -15,49 +15,62 @@ type RemovePrefix<
   ? T
   : T["length"] extends 0
     ? PartialBindError<"Too many args provided to function">
-    : T extends [infer THead, ...infer TRest]
-      ? Prefix extends [infer PrefixHead, ...infer PrefixRest]
-        ? PrefixHead extends THead
+    : T extends readonly [infer THead, ...infer TRest]
+      ? Prefix extends readonly [infer PrefixHead, ...infer PrefixRest]
+        ? // Both T and Prefix have a non-rest head.
+          PrefixHead extends THead
           ? RemovePrefix<TRest, PrefixRest>
           : PartialBindError<"Argument of the wrong type provided to function">
-        : Prefix extends Array<THead>
-          ? [THead?, ...RemovePrefix<TRest, Prefix>]
-          : never
-      : T extends [(infer THead)?, ...infer TRest]
-        ? Prefix extends [infer PrefixHead, ...infer PrefixRest]
-          ? PrefixHead extends THead
+        : Prefix extends ReadonlyArray<THead>
+          ? // T has a non-rest head, Prefix is an array.
+            // Prefix could possibly be empty, so this has to be THead?.
+            [THead?, ...RemovePrefix<TRest, Prefix>]
+          : // Prefix is e.g. [...Array<number>, string]. We can't do a
+            // type-level prefix removal, so we return an error.
+            PartialBindError<"Can't infer type of provided args">
+      : // T has an optional or rest parameter head. If T is a parameter list,
+        // this can only happen if we have optional arguments or a rest param;
+        // both cases are similar.
+        T extends readonly [(infer THead)?, ...infer TRest]
+        ? Prefix extends readonly [infer PrefixHead, ...infer PrefixRest]
+          ? // Prefix has a non-rest head.
+            PrefixHead extends THead
             ? RemovePrefix<TRest, PrefixRest>
             : PartialBindError<"Argument of the wrong type provided to function">
-          : Prefix extends TRest
-            ? TRest
-            : never
-        : never;
+          : Prefix extends readonly [THead?, ...TRest]
+            ? // Prefix is an array. It must match *both* the optional type
+              // and the rest param.
+              TRest
+            : PartialBindError<"Argument of the wrong type provided to function">
+        : // We got passed a parameter list that isn't what we expected; this
+          // is an internal error.
+          PartialBindError<1>;
 
 type PartiallyBound<
   F extends (...args: any) => any,
-  T extends IterableContainer,
-> = (...rest: RemovePrefix<Parameters<F>, T>) => ReturnType<F>;
+  Partial extends IterableContainer,
```

Comment:
Don't use partial as the name because it's also a built-in utility type, which can be confusing (and will break grepping the code for Partial).